### PR TITLE
Add explicit dynamic fusion arg

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72077
* #72036
* #72038
* #72040
* #72032
* #71651
* #71650
* __->__ #71173

Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

### CODE REVIEW

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 1]
```diff
@@ -1236,7 +1249,8 @@ void FuseTensorExprs(
 }
 
 Operation createTensorExprOp(const Node* node) {
-  if (!tensorExprDynamicShapeFusionEnabled()) {
+  bool dynamic_shape_fusion_node = node->hasAttribute(attr::striding_inputs_desc);
```
Comment (navahgar) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:1252]: Is every `TensorExprGroup` node with dynamic shapes guaranteed to have `attr::striding_inputs_desc`? What if there are no dynamic strides?

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 2]
```diff
@@ -1210,23 +1216,30 @@ class TensorExprFuser {
...
+    bool fuse_to_dynamic_shapes) {
   GRAPH_DUMP("Before TExprFuser: ", graph);
 
   // Temporary change for Block code generation.
   if (tensorexpr::getTEGenerateBlockCode()) {
     min_group_size = 1;
   }
 
+  if (add_composed_op) {
+    TORCH_INTERNAL_ASSERT(fuse_to_dynamic_shapes, "Fusing static shapes with composed op NYI");
```
Comment (navahgar) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:1236]: Why is this an error?

**Review by navahgar — COMMENTED** [review]
Mostly LGTM. Couple of minor comments.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1236]: The `add_composed_op` path always fuses to dynamic shapes right now, we didn't implement composed op + static shapes (bc there is no use case for it)

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1252]: yes always guaranteed. If there are no dynamic strides then there will be `attr::striding_inputs_desc` with TENSOR_CONT description

**Review by navahgar — APPROVED** [review]
LGTM

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

**Comment by xsacha** [comment]
This commit appears to have broken some of my models at specific input sizes.
See #72173

**Pull request closed without merging**